### DISCUSSION TITLE
How to create tokens for conference rooms

### DISCUSSION THREAD

**Discussion by Mohit-3196** [ant-media/Ant-Media-Server#5893] [FAQs]
Hi,
Is there a way to create tokens for the conference rooms as we do for the streams so that the rooms can be secured as well?
The typical scenarios are like
- [ ] When a user has a publish token for the room, he/she can publish a stream to the room.
- [ ] When a user has a play token for the room, he/she can play any stream in that room. There should be no separate token for playing each stream.

Thank you

**Comment by mekya** [comment]
Hi @Mohit-3196,

Sorry for responding late. Let me tell the status here

In v2.7.0, users can only use TOTP tokens to publish and play the conference with the same token. In next relese(2.8.0), all tokens(JWT, One-Time, Hash-based) will be used in the same way. It can be tested in the snapshot version. 

The magic is that Publish tokens generated for a conference room id (aka. streamId) can be used to play the streams as well. Moreover, in TOTP, there is no need to register the subscribers in advance.  After this context information, let me tell how to do that in step by step

1. Go to the web panel and enable TOTP for publish and play as below

<img width="1711" alt="Screenshot 2023-12-24 at 12 13 04" src="https://github.com/ant-media/Ant-Media-Server/assets/3456251/cd469126-685b-4506-9858-764fa6ac8c8b">

2. Assume that you're going to join the room with having streamId `room1` with subscriberId `myUsername`. Then generate the TOTP token to publish the room with the [following TOTP REST method](https://antmedia.io/rest/#/BroadcastRestService/getTOTP)

   Check the sample below for `room1` and subscriber `myUsername`.  Please change the values according to your case. 
```
  curl -X 'GET' 
  'https://{YOUR_SERVER_ADDRESS}:5443/{YOUR_APPLICATION}/rest/v2/broadcasts/room1/subscribers/myUsername/totp?type=publish'
```
3. Use the subscriberId `myUsername` and returned code as `subscriberCode` from the REST method in SDKs to publish and play the streams in `room1` 



Lastly, if you want to try the latest snapshot for using JWT, One-Time etc. to publish and play the same room with single token, you can download the latest version from your account at antmedia.io. It will be listed as `Ant Media Server Latest Snapshot(Unstable)`


Please let me know if there is something missing and it does not work for you.

**Reply by Mohit-3196** [reply]
Thank you @mekya abi.
Much appreciated :)
